graphql-server/pkg/dataloaders/proposal.go

```diff
@@ -60,9 +80,89 @@ func NewProposalDataloader(proposalQuery queries.IProposalQuery) ProposalDataloa
 		Wait:     DefaultWait,
 	})
 
+	proposalVoteLoader := godataloader.NewDataLoader(godataloader.DataLoaderConfig[ProposalVoteKey, *models.ProposalVote]{
+		MaxBatch: DefaultMaxBatch,
+		Wait:     DefaultWait,
+		Fetch: func(keys []ProposalVoteKey) ([]*models.ProposalVote, []error) {
+			proposalIDs := []int{}
+			addresses := []string{}
+			for _, key := range keys {
+				if !slices.Contains(proposalIDs, key.ProposalID) {
+					proposalIDs = append(proposalIDs, key.ProposalID)
+				}
+				if !slices.Contains(addresses, key.Address) {
+					addresses = append(addresses, key.Address)
+				}
+			}
+
+			votes, err := proposalQuery.QueryProposalVotes(proposalIDs, addresses)
+			if err != nil {
+				errors := make([]error, 0, len(keys))
+				for range keys {
+					errors = append(errors, err)
+				}
+				return nil, errors
+			}
+
+			voteKeyToVote := make(map[ProposalVoteKey]*models.ProposalVote, len(votes))
+			for _, vote := range votes {
+				key := ProposalVoteKey{vote.ProposalID, vote.VoterAddress}
+				voteKeyToVote[key] = vote
+			}
+
+			// map query results to input keys' order
+			res := make([]*models.ProposalVote, len(keys))
+			for index, key := range keys {
+				res[index] = voteKeyToVote[key]
+			}
+			return res, nil
+		},
+	})
+
+	proposalDepositLoader := godataloader.NewDataLoader(godataloader.DataLoaderConfig[ProposalDepositKey, *models.ProposalDeposit]{
+		MaxBatch: DefaultMaxBatch,
+		Wait:     DefaultWait,
+		Fetch: func(keys []ProposalDepositKey) ([]*models.ProposalDeposit, []error) {
+			proposalIDs := []int{}
+			addresses := []string{}
+			for _, key := range keys {
+				if !slices.Contains(proposalIDs, key.ProposalID) {
```

Comment:
this is slow, can you tell me the big-o notation for this?

and what would be a better way to do this?

Reply:
> this is slow, can you tell me the big-o notation for this?

o(n^2) 

> and what would be a better way to do this?

as discussed offline, will use a map here instead

Reply:
in particular, we can use `map[T]struct{}` in go as a set of T (since there is no native set type in go)